python-checks/src/main/java/org/sonar/python/checks/IdentityComparisonWithCachedTypesCheck.java

```diff
@@ -0,0 +1,131 @@
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.IsExpression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.Token;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.types.InferredType;
+import org.sonar.python.tree.TreeUtils;
+
+import static java.util.Arrays.asList;
+
+// https://jira.sonarsource.com/browse/RSPEC-5795
+// https://jira.sonarsource.com/browse/SONARPY-673
+@Rule(key = "S5795")
+public class IdentityComparisonWithCachedTypesCheck extends PythonSubscriptionCheck {
+  private static final String MESSAGE_IS = "Replace this \"is\" operator with \"==\"; identity operator is not reliable here.";
+  private static final String MESSAGE_IS_NOT = "Replace this \"is not\" operator with \"!=\"; identity operator is not reliable here.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.IS, IdentityComparisonWithCachedTypesCheck::checkIsComparison);
+  }
+
+  private static void checkIsComparison(SubscriptionContext subscriptionContext) {
+    IsExpression isExpr = (IsExpression) subscriptionContext.syntaxNode();
+    if (isUnsuitableOperand(isExpr.leftOperand()) || isUnsuitableOperand(isExpr.rightOperand())) {
+      Token notToken = isExpr.notToken();
+      if (notToken == null) {
+        subscriptionContext.addIssue(isExpr.operator(), MESSAGE_IS);
+      } else {
+        subscriptionContext.addIssue(isExpr.operator(), notToken, MESSAGE_IS_NOT);
+      }
+    }
+  }
+
+  /**
+   * Fully qualified names of constructors and functions that would create values unsuitable for
+   * comparison with <code>is</code>.
+   *
+   * Note that these are fully qualified names of (value-level) expressions, not types.
+   */
+  private static final Set<String> VULNERABLE_CONSTRUCTOR_FQNS = new HashSet<>(
+    asList("frozenset", "bytes", "int", "float", "str", "tuple", "hash"));
+
+  /**
+   * Names of types that usually should not be compared with <code>is</code>.
+   *
+   * Note that these are names of types, not `fqn`s of expressions.
+   */
+  private static final Set<String> NAMES_OF_TYPES_UNSUITABLE_FOR_COMPARISON = new HashSet<>(
+    asList("frozenset", "bytes", "int", "float", "tuple"));
+
+  private static boolean isUnsuitableOperand(Expression expr) {
+    InferredType tpe = expr.type();
+    if (isUnsuitableType(tpe)) {
+      if (expr.is(Tree.Kind.NAME)) {
+        Symbol symb = ((Name) expr).symbol();
+        // Impossible to cover. If the type could be inferred, then the `symb` cannot be null.
+        return symb != null && !isVariableThatCanEscape(symb);
+      } else {
+        return true;
+      }
+    }
+    if (expr.is(Tree.Kind.STRING_LITERAL)) {
+      return true;
+    } else if (expr.is(Tree.Kind.CALL_EXPR)) {
+      Symbol calleeSymbol = ((CallExpression) expr).calleeSymbol();
```

Comment:
As discussed, I think this deserves more explanation about why don't we just rely on types.